Python/cpuinfo.c

```diff
@@ -0,0 +1,556 @@
+ */
+#define CPUID_APPLY_MACRO(MACRO)        \
+    do {                                \
+        MACRO(sse);                     \
+        MACRO(sse2);                    \
+        MACRO(sse3);                    \
+        MACRO(ssse3);                   \
+        MACRO(sse41);                   \
+        MACRO(sse42);                   \
+                                        \
+        MACRO(avx);                     \
+        MACRO(avx_ifma);                \
+        MACRO(avx_ne_convert);          \
+                                        \
+        MACRO(avx_vnni);                \
+        MACRO(avx_vnni_int8);           \
+        MACRO(avx_vnni_int16);          \
+                                        \
+        MACRO(avx2);                    \
+                                        \
+        MACRO(avx512_f);                \
+        MACRO(avx512_cd);               \
+                                        \
+        MACRO(avx512_er);               \
+        MACRO(avx512_pf);               \
+                                        \
+        MACRO(avx512_4fmaps);           \
+        MACRO(avx512_4vnniw);           \
+                                        \
+        MACRO(avx512_vpopcntdq);        \
+                                        \
+        MACRO(avx512_vl);               \
+        MACRO(avx512_dq);               \
+        MACRO(avx512_bw);               \
+                                        \
+        MACRO(avx512_ifma);             \
+        MACRO(avx512_vbmi);             \
+                                        \
+        MACRO(avx512_vnni);             \
+                                        \
+        MACRO(avx512_vbmi2);            \
+        MACRO(avx512_bitalg);           \
+                                        \
+        MACRO(avx512_vp2intersect);     \
+                                        \
+        MACRO(cmov);                    \
+        MACRO(fma);                     \
+        MACRO(popcnt);                  \
+        MACRO(pclmulqdq);               \
+                                        \
+        MACRO(xsave);                   \
+        MACRO(osxsave);                 \
+                                        \
+        MACRO(xcr0_sse);                \
+        MACRO(xcr0_avx);                \
+        MACRO(xcr0_avx512_opmask);      \
+        MACRO(xcr0_avx512_zmm_hi256);   \
+        MACRO(xcr0_avx512_hi16_zmm);    \
+    } while (0)
+
+void
+_Py_cpuid_disable_features(py_cpuid_features *flags)
+{
+    flags->maxleaf = 0;
+#define CPUID_DISABLE(FLAG)    flags->FLAG = 0
+    CPUID_APPLY_MACRO(CPUID_DISABLE);
+#undef CPUID_DISABLE
+}
+
+int
+_Py_cpuid_has_features(const py_cpuid_features *actual,
+                       const py_cpuid_features *expect)
+{
+    if (!actual->ready || !expect->ready) {
+        return 0;
+    }
+    if (actual->maxleaf < expect->maxleaf) {
+        return 0;
+    }
+#define CPUID_CHECK_FEATURE(FLAG)               \
```

Comment:
This is also over-encapsulation, let's remove it.
(This suggestion applies to macros that are local to functions only.)